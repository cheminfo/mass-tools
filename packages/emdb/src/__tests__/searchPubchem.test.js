'use strict';

const searchPubchem = require('../searchPubchem.js');

jest.setTimeout(50000);
describe('searchPubchem', () => {
  it('simple case', async () => {
    let data = await searchPubchem(60, {
      ionizations: '(H+)2, H+',
      precision: 10,
      url: 'https://pubchem-beta.cheminfo.org/mfs/v1/fromEM',
    });
    expect(data).toHaveLength(2);
    expect(data).toMatchSnapshot();
  });

  it('string containing more than 1 monoisotopic mass', async () => {
    let data = await searchPubchem('12,24,36', {
      ionizations: '',
      precision: 10,
      minPubchemEntries: 0,
      url: 'https://pubchem-beta.cheminfo.org/mfs/v1/fromEM',
    });
    expect(data).toHaveLength(3);
    expect(data).toMatchSnapshot();
  });

  it('simple case negative ionization', async () => {
    let data = await searchPubchem(60, {
      ionizations: '(H+)-2, (H+)-1',
      precision: 10,
      url: 'https://pubchem-beta.cheminfo.org/mfs/v1/fromEM',
    });
    expect(data).toHaveLength(3);
    expect(data).toMatchSnapshot();
  });

  it('highly precise', async () => {
    let data = await searchPubchem(81.06987671016094, {
      ionizations: 'H+',
      precision: 1,
      url: 'https://pubchem-beta.cheminfo.org/mfs/v1/fromEM',
    });
    expect(data).toHaveLength(1);
  });

  it('simple case with range filter', async () => {
    let data = await searchPubchem(60, {
      ionizations: '(H+)2, H+',
      precision: 100,
      ranges: 'C0-10H0-10N0-10O0-10',
      url: 'https://pubchem-beta.cheminfo.org/mfs/v1/fromEM',
    });
    expect(data).toHaveLength(1);
  });
});
