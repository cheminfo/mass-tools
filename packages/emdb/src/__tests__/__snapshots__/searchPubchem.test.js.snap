// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchPubchem simple case 1`] = `
[
  {
    "atoms": {
      "C": 2,
      "H": 4,
      "N": 2,
      "P": 2,
    },
    "charge": 0,
    "em": 117.98497213462,
    "info": {
      "nbPubchemEntries": 7,
    },
    "ionization": {
      "atoms": {
        "H": 2,
      },
      "charge": 2,
      "em": 2.01565006446,
      "mf": "(H+)2",
    },
    "mass": 118.0141652294252,
    "mf": "C2H4N2P2",
    "monoisotopicMass": 117.98497213462,
    "ms": {
      "charge": 2,
      "delta": 0.00023748036907278447,
      "em": 59.99976251963093,
      "ionization": "(H+)2",
      "ppm": 3.958021817087773,
    },
    "unsaturation": 3,
  },
  {
    "atoms": {
      "C": 1,
      "H": 2,
      "N": 1,
      "P": 1,
    },
    "charge": 0,
    "em": 58.99248606731,
    "info": {
      "nbPubchemEntries": 9,
    },
    "ionization": {
      "atoms": {
        "H": 1,
      },
      "charge": 1,
      "em": 1.00782503223,
      "mf": "H+",
    },
    "mass": 59.0070826147126,
    "mf": "CH2NP",
    "monoisotopicMass": 58.99248606731,
    "ms": {
      "charge": 1,
      "delta": 0.00023748036907278447,
      "em": 59.99976251963093,
      "ionization": "H+",
      "ppm": 3.958021817087773,
    },
    "unsaturation": 2,
  },
]
`;

exports[`searchPubchem simple case negative ionization 1`] = `
[
  {
    "atoms": {
      "C": 2,
      "H": 6,
      "N": 2,
      "O": 2,
      "S": 1,
    },
    "charge": 0,
    "em": 122.01499861577999,
    "info": {
      "nbPubchemEntries": 69,
    },
    "ionization": {
      "atoms": {
        "H": -2,
      },
      "charge": -2,
      "em": -2.01565006446,
      "mf": "(H+)-2",
    },
    "mass": 122.14611999546037,
    "mf": "C2H6N2O2S",
    "monoisotopicMass": 122.01499861577999,
    "ms": {
      "charge": -2,
      "delta": -0.00022285556906354032,
      "em": 60.00022285556906,
      "ionization": "(H+)-2",
      "ppm": -3.714245688720062,
    },
    "unsaturation": 1,
  },
  {
    "atoms": {
      "C": 4,
      "Cl": 1,
      "H": 7,
      "O": 2,
    },
    "charge": 0,
    "em": 122.01345714675,
    "info": {
      "nbPubchemEntries": 155,
    },
    "ionization": {
      "atoms": {
        "H": -2,
      },
      "charge": -2,
      "em": -2.01565006446,
      "mf": "(H+)-2",
    },
    "mass": 122.55027629657599,
    "mf": "C4H7ClO2",
    "monoisotopicMass": 122.01345714675,
    "ms": {
      "charge": -2,
      "delta": 0.0005478789459303357,
      "em": 59.99945212105407,
      "ionization": "(H+)-2",
      "ppm": 9.131399147194589,
    },
    "unsaturation": 1,
  },
  {
    "atoms": {
      "C": 5,
      "H": 1,
    },
    "charge": 0,
    "em": 61.00782503223,
    "info": {
      "nbPubchemEntries": 9,
    },
    "ionization": {
      "atoms": {
        "H": -1,
      },
      "charge": -1,
      "em": -1.00782503223,
      "mf": "(H+)-1",
    },
    "mass": 61.061620237732015,
    "mf": "C5H",
    "monoisotopicMass": 61.00782503223,
    "ms": {
      "charge": -1,
      "delta": -0.0005485799090720889,
      "em": 60.00054857990907,
      "ionization": "(H+)-1",
      "ppm": -9.142914890877822,
    },
    "unsaturation": 5.5,
  },
]
`;

exports[`searchPubchem string containing more than 1 monoisotopic mass 1`] = `
[
  {
    "atoms": {
      "C": 1,
    },
    "charge": 0,
    "em": 12,
    "info": {
      "nbPubchemEntries": 1,
    },
    "ionization": {
      "atoms": {},
      "charge": 0,
      "em": 0,
      "mf": "",
    },
    "mass": 12.010735896735248,
    "mf": "C",
    "monoisotopicMass": 12,
    "ms": {
      "charge": 0,
      "delta": NaN,
      "em": 0,
      "ionization": "",
      "ppm": NaN,
    },
    "unsaturation": 2,
  },
  {
    "atoms": {
      "C": 2,
    },
    "charge": 0,
    "em": 24,
    "info": {
      "nbPubchemEntries": 1,
    },
    "ionization": undefined,
    "mass": 24.021471793470496,
    "mf": "C2",
    "monoisotopicMass": 24,
    "ms": {
      "charge": 0,
      "delta": NaN,
      "em": 0,
      "ionization": "",
      "ppm": NaN,
    },
    "unsaturation": 3,
  },
  {
    "atoms": {
      "C": 3,
    },
    "charge": 0,
    "em": 36,
    "info": {
      "nbPubchemEntries": 2,
    },
    "ionization": undefined,
    "mass": 36.032207690205745,
    "mf": "C3",
    "monoisotopicMass": 36,
    "ms": {
      "charge": 0,
      "delta": NaN,
      "em": 0,
      "ionization": "",
      "ppm": NaN,
    },
    "unsaturation": 4,
  },
]
`;
